src/Tools/dotnet-monitor/EgressService.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring;
+using Microsoft.Diagnostics.Monitoring.Egress;
+using Microsoft.Diagnostics.Monitoring.Egress.AzureStorage;
+using Microsoft.Diagnostics.Monitoring.Egress.FileSystem;
+using Microsoft.Extensions.Options;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    internal class EgressService : IEgressService
+    {
+        private readonly IOptionsMonitor<EgressOptions> _egressOptions;
+        private readonly IOptionsMonitor<AzureStorageOptions> _azureStorageOptions;
+
+        public EgressService(
+            IOptionsMonitor<EgressOptions> egressOptions,
+            IOptionsMonitor<AzureStorageOptions> azureStorageOptions)
+        {
+            _egressOptions = egressOptions;
+            _azureStorageOptions = azureStorageOptions;
+        }
+
+        public Task<EgressResult> EgressAsync(string endpointName, Func<CancellationToken, Task<Stream>> action, string fileName, string contentType, IEndpointInfo source, CancellationToken token)
+        {
+            if (_egressOptions.CurrentValue.Endpoints.TryGetValue(endpointName, out EgressEndpointOptions template))
+            {
+                switch (template)
+                {
+                    case AzureBlobEgressEndpointOptions azureBlobSettings:
```

Comment:
My approach was to create an egress infrastructure that was not inherently tied to dotnet-monitor and could be usable by other libraries and executables without forcing them to understand conventions that were established by the needs of dotnet-monitor. I also did not want the egress infrastructure to be forced into knowing that it is used by the REST API (so no dependencies on ASP.NET Core).

For example, The EndpointType enum found in the ServiceCollectionExtensions class in the dotnet-monitor project is an implementation specific to how dotnet-monitor describes the endpoint information (the egress implementations and abstractions do not know of the "type" property of the configuration section that contains the endpoint options) so that the name "AzureBlobStorage" or "FileSystem" are not inherently tied to the endpoints and options classes.

Because the the EgressService and ServiceCollectionExtensions classes are implementation details of the dotnet-monitor project, I don't see them agreeing on how to parse data from the configuration sections and how to fill in details as anything more than that's just how dotnet-monitor does it. The latter is responsible for parsing the options out of the configuration section. The EgressService has the helper methods because it needs to compose data from three sources: (1) the Egresss:Endpoints configuration, (2) the AzureStorage:SasTokens configuration, and (3) whatever additional options that the caller needs to provide as part of egressing the actual data (e.g. content type and blob metadata). The additional stream specific options need to be a separate implementation object rather than part of the egress API so that each endpoint type can have its own options.

The code write up you have suggests to me that you may be viewing these egress concepts as implementation details of dotnet-monitor rather than a new set of APIs that could be used by other libraries and executables. The TryParseSection methods would have to know dotnet-monitor details in order to parse the data (the name of the configuration section containing the endpoint options, the way to parse the endpoint type, and configuration section containing the Azure storage SAS tokens). The Azure blob egress provider would have to know what data to collect from the IEndpointInfo (and probably from the environment block of the process, such as the host name (which is the pod name) if running in k8s) in order to populate metadata on the blob. Maybe these providers are wrapper abstractions for the egress concepts as they currently exist in the PR, but translate the dotnet-monitor implementation details to the egress API?

Reply:
I've now rewritten the configuration of the EgressOptions to use the provider notion. Each egress type is registered as a provider. The provider knows how to parse the options from the configuration section and produce what I'm currently calling a ConfiguredEgressEndpoint. The class allows egress invocation without exposing the implementation details of the specific egress endpoint. The EgressEndpoint implementations largely remain unchanged at this time.